File: diag.cpp
//TC: O(m*n) where m is number of rows and n is number of columns
//SC: O(1) 

class Solution {
public:
    vector<int> findDiagonalOrder(vector<vector<int>>& matrix) {
        
        vector<int> result;
            
        if(matrix.size() == 0) return result;
        int dir = 1; //1=up, -1=down
        int r = 0; 
        int c = 0;
        int m = matrix.size(); //num_rows
        int n = matrix[0].size(); //num_cols
                
        while(result.size()<m*n){
            result.push_back(matrix[r][c]);
            if(dir == 1){
                //check if column goes out of bounds
                //if column is last, then increment row, change direction
                if(c == n-1){
                    r = r+1;
                    dir = -1;
                }
                else if(r==0){
                    c++;
                    dir = -1;
                }
                else {
                     r--;
                     c++;
                }
            }
            
            else {
                //if row is last, increment column, change direction
                if(r == m-1){
                    c = c+1;
                    dir = 1;
                }
                else if(c==0){
                    r++;
                    dir = 1;
                }
                else {
                     c--;
                     r++;
                }                
            }       
        }
    
    return result;
        
        
    }
};

File: diag.py
"""498. Diagonal Traverse
Time Complexity: 0(NM)
Space Complexity: O(1)
"""
class Solution:
    def findDiagonalOrder(self, mat: List[List[int]]) -> List[int]:
        if len(mat)==0:
            return []
        rows = len(mat)
        cols = len(mat[0])
        res_arr = [None]*(rows*cols)
        
        index = 0
        matrix_r = 0
        matrix_c = 0
        
        direction = 1 #going up
        while(index<(rows*cols)):
            res_arr[index] = mat[matrix_r][matrix_c]
            if direction ==1:
                if matrix_c+1>cols-1:
                    matrix_r+=1 #reached 2
                    direction = 0 #now we go down
                elif matrix_r-1 <0: ##reached 1
                    matrix_c+=1
                    direction=0
                else:             #all other case
                    matrix_r-=1
                    matrix_c+=1   
            else: # going down
                if matrix_r+1>rows-1: # last row - now we need t go up
                    matrix_c+=1
                    direction =1
                elif matrix_c -1<0:
                    matrix_r+=1
                    direction =1
                else:
                    matrix_r+=1
                    matrix_c-=1     
            index+=1
        return res_arr
                
            